test/models/maintenance_tasks/default_task_loader_test.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require "test_helper"
+
+module MaintenanceTasks
+  class DefaultTaskLoaderTest < ActiveSupport::TestCase
+    include ActiveSupport::Testing::Isolation
+
+    test ".load_all loads all autoloadable Task subclasses within the Maintenance namespace" do
```

Comment:
This new test fails in Rails 6.
Has it possibly identified an issue with the gem in older Rails versions? I haven't altered the recursive loading algorithm, just moved it from `Task.load_constants` to `DefaultTaskLoader.load_constants`. 🤔

Reply:
```
Failure:
MaintenanceTasks::DefaultTaskLoaderTest#test_.load_all_loads_all_autoloadable_Task_subclasses_within_the_Maintenance_namespace [/home/runner/work/maintenance_tasks/maintenance_tasks/vendor/bundle/ruby/3.0.0/gems/activesupport-6.0.6.1/lib/active_support/inflector/methods.rb:284]:
unexpected invocation: Maintenance.const_get("Nested")
unsatisfied expectations:
- expected exactly once, invoked never: Maintenance.const_get(:Nested)
- expected exactly once, invoked never: Maintenance.const_get(:ImportPostsTask)
- expected exactly once, invoked never: Maintenance.const_get(:ErrorTask)
- expected exactly once, invoked never: Maintenance.const_get(:EnqueueErrorTask)
- expected exactly once, invoked never: Maintenance.const_get(:CancelledEnqueueTask)
- expected exactly once, invoked never: Maintenance.const_get(:CallbackTestTask)
- expected exactly once, invoked never: Maintenance.const_get(:BatchImportPostsTask)
```

Wonder if it's just because of a different order... 🤔

Reply:
No, the order returned by `.constants` seems to differ depending on the version of Rails.
This would work in Ruby 6.1 for example:

```diff
diff --git i/app/models/maintenance_tasks/default_task_loader.rb w/app/models/maintenance_tasks/default_task_loader.rb
index 14f1624..fed0ea5 100644
--- i/app/models/maintenance_tasks/default_task_loader.rb
+++ w/app/models/maintenance_tasks/default_task_loader.rb
@@ -15,7 +15,8 @@ module MaintenanceTasks
       private
 
       def load_constants(namespace)
-        namespace.constants.each do |name|
+        namespace.constants.sort.each do |name|
+          puts "#{namespace}.const_get(#{name})"
           object = namespace.const_get(name)
           load_constants(object) if object.instance_of?(Module)
         end
diff --git i/test/models/maintenance_tasks/default_task_loader_test.rb w/test/models/maintenance_tasks/default_task_loader_test.rb
index 8fd8ef9..b029518 100644
--- i/test/models/maintenance_tasks/default_task_loader_test.rb
+++ w/test/models/maintenance_tasks/default_task_loader_test.rb
@@ -8,6 +8,9 @@ module MaintenanceTasks
 
     test ".load_all loads all autoloadable Task subclasses within the Maintenance namespace" do
       [
+        Maintenance::Nested::NestedMore::NestedMoreTask,
+        Maintenance::Nested::NestedTask,
+        Maintenance::Nested::NestedMore,
         Maintenance::BatchImportPostsTask,
         Maintenance::CallbackTestTask,
         Maintenance::CancelledEnqueueTask,
@@ -15,9 +18,6 @@ module MaintenanceTasks
         Maintenance::ErrorTask,
         Maintenance::ImportPostsTask,
         Maintenance::Nested,
-        Maintenance::Nested::NestedMore,
-        Maintenance::Nested::NestedMore::NestedMoreTask,
-        Maintenance::Nested::NestedTask,
         Maintenance::NoCollectionTask,
         Maintenance::ParamsTask,
         Maintenance::TestTask,
```

Reply:
Actually @davidstosik if I change the order to what you proposed in https://github.com/Shopify/maintenance_tasks/pull/867#discussion_r1304083757, it passes against all the Rails versions for me? Were you seeing failures locally?

Reply:
@adrianna Huh, wow, this is weird, you seem to be right. Although both orders work with Rails 7, it seems only a specific order works with Rails 6.1?

For fun (and because I hadn't imagined the right order would work for all supported Rails versions), I've also implemented a spy. Here it is, just to show what it could look like:

```rb
# frozen_string_literal: true

require "test_helper"

module MaintenanceTasks
  class DefaultTaskLoaderTest < ActiveSupport::TestCase
    class ConstGetSpy < SimpleDelegator
      cattr_reader :calls, default: Hash.new { |h, k| h[k] = [] }

      def const_get(name)
        calls[spied_module] << name

        return_value = super

        if return_value.is_a?(Module)
          self.class.new(return_value)
        else
          return_value
        end
      end

      def instance_of?(klass)
        spied_module.instance_of?(klass)
      end

      private

      def spied_module
        __getobj__
      end
    end

    SPIED_MAINTENANCE_TASKS_MODULE = ConstGetSpy.new(Maintenance)

    test ".load_all loads all autoloadable Task subclasses within the Maintenance namespace" do
      original_tasks_module = MaintenanceTasks.tasks_module
      MaintenanceTasks.tasks_module = "MaintenanceTasks::DefaultTaskLoaderTest::SPIED_MAINTENANCE_TASKS_MODULE"

      expected_calls_to_const_get = {
        Maintenance => [
          :BatchImportPostsTask,
          :CallbackTestTask,
          :CancelledEnqueueTask,
          :EnqueueErrorTask,
          :ErrorTask,
          :ImportPostsTask,
          :Nested,
          :NoCollectionTask,
          :ParamsTask,
          :TestTask,
          :UpdatePostsInBatchesTask,
          :UpdatePostsModulePrependedTask,
          :UpdatePostsTask,
          :UpdatePostsThrottledTask,
        ],
        Maintenance::Nested => [
          :NestedMore,
          :NestedTask,
        ],
        Maintenance::Nested::NestedMore => [
          :NestedMoreTask,
        ]
      }

      DefaultTaskLoader.load_all

      assert_equal(expected_calls_to_const_get, ConstGetSpy.calls.transform_values(&:sort))

      MaintenanceTasks.tasks_module = original_tasks_module
    end
  end
end
```

(I saw your more general comment about the PR, will reply next. 🙇🏻‍♂️ )